app/grandchallenge/components/serializers.py

```diff
@@ -172,3 +181,72 @@ class HyperlinkedComponentInterfaceValueSerializer(
         read_only=True,
         allow_null=True,
     )
+
+
+class CIVSetPostSerializerMixin:
+
+    editability_error_message = "This object cannot be updated."
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.fields["values"] = ComponentInterfaceValuePostSerializer(
+            many=True,
+            context=self.context,
+            required=False,
+        )
+
+    def create(self, validated_data):
+        if validated_data.pop("values", None):
+            raise DRFValidationError("Values can only be added via update")
+        return super().create(validated_data)
+
+    def update(self, instance, validated_data):
+
+        if not instance.is_editable:
+            raise DRFValidationError(self.editability_error_message)
+
+        values = validated_data.pop("values", [])
+
+        instance = super().update(instance, validated_data)
+
+        request = self.context["request"]
+
+        civs = []
+
+        for value in values:
+            interface = value["interface"]
+            upload_session = value.get("upload_session", None)
+            user_upload = value.get("user_upload", None)
+            image = value.get("image", None)
+            value = value.get("value", None)
+            civs.append(
+                CIVData(
+                    interface_slug=interface.slug,
+                    value=upload_session or user_upload or image or value,
+                )
+            )
+        try:
+            instance.validate_values_and_execute_linked_task(
+                values=civs,
+                user=request.user,
+            )
+        except CIVNotEditableException as e:
+            error_handler = instance.get_error_handler()
+            error_handler.handle_error(
+                error_message="An unexpected error occurred",
+                user=request.user,
+            )
+            logger.error(e, exc_info=True)
+
+        if not self.partial:
+            instance.refresh_from_db()
+            current_civs = {
+                civ.interface.slug: civ for civ in instance.values.all()
+            }
+            current_interfaces = set(current_civs.keys())
+            updated_interfaces = {v["interface"].slug for v in values}
+            for interface in current_interfaces - updated_interfaces:
+                self.instance.remove_civ(civ=current_civs[interface])
```

Comment:
If this happens for image interfaces, the user will loose permission to use the image. I remember a discussion with James that if we allow removal of images from display sets or archive items, that they need to be moved to a different AI or DS first, so that the user doesn't loose access, see [here](https://github.com/comic/grand-challenge.org/pull/3156#issuecomment-1884507499). I'm not sure if this is still a requirement, but thought I'd mention it. It feels a bit like a strange side-effect to create a new DS/AI for the removed image, especially given that users can then remove the AI/DS and lose access anyway. Whichever way we go, I think the consequences need to be clear to the user (either losing access or creating a new AI/DS), I think.

Reply:
Huh? The idea here is that replacing an image-linked CIV within the same CIVSet works because the new CIV is first created and the old one is only later removed. This should keep the linked image bound to the CIVSet. Much what James is warning against.

From your description, it seems you suggest to move it to a completely separate CIVSet? That seems to go against the idea of PUT (or REST at that). 

If users decide to run a (somewhat weird) update workflow against a reader-study or archive where they locally store a reference to an image, then PUT completely new values to a CIVSet and then decide to create a new CIV off of the image ref only to find it's no longer accessible... that is really their problem. We can only guarantee things to stay cohesive in the scope of a single request.






with an image linked gets used by a different interface (with the same CIVSet) the new CIV

Reply:
This is a problem but I think the solution of creating a new archive item to hold is was not the correct one (rather, the `view_image` permission should have been attached through the upload session too). I don't think we need to maintain that solution if we still have it - I thought that we removed it?

Reply:
Ah true, we no longer do that. When someone uploads a new image to an archive item, we don't create a new item with the previous image, same for DS. We used to do that, at least for display sets, but that was removed indeed.

Reply:
Not 100% sure, does this PR pivot on the view_image permissions that still need to be added via user upload?

Reply:
Not sure either, here is how it works right now: We do not attach view permissions to upload sessions at the moment. Image viewing permissions are set through the objects the images are attached to (job, display sets, archive items, rs answers). The piece of code that handles the logic is `update_viewer_groups_permissions`. This gets called when an archive item or display set get edited, for example. 

I didn't mean to suggest that creating a new CIV set is the right/best solution, I just mentioned this because that is how we used to deal with the issue of losing access . Attaching permissions to the upload session would be better - these persist forever, so we don't delete upload session, like we do with user uploads, right? 

But I think that change requires quite some changes in the codebase, so not something we would do in a cooldown week PR, IMO. Is this a blocker for this PR? I actually don't think so since we anyway already removed the creation of new CIVsets, so users are already losing access to their images if they update the DS or AI... but maybe James can comment?

Reply:
It wouldn't be that much of a migration to be honest - Images already have user "view_image" permissions for answer creators. It gets in to an area where we have to think about behaviour though: is the user who uploads the image really the owner of the data? Presumably the user is currently working for an organisation, and they're uploading the organisations data, in which case they should probably loose access to the data if they leave their job. Likewise - you could argue this is the case for any other object in GC (Challenges, Archives, Reader Studies) - so the usual way to handle this is that they forfeit their account when they leave the organisation, just like they forfeit their email address.

So, I don't know what to do here. Users already lose access when replacing, no one has actually complained about it AFAIK, so I don't see a user-need to change the behaviour.

Reply:
Agreed. So then this can be merged.